### ISSUE TITLE
Dynamic Severity checking

### ISSUE THREAD

**Issue by vlatkoB** [Soostone/katip#101]
Hi,
I've implemented dynamic LogEnv and would like to allow changing `Severity` on all active `Scribes`.
Currently, the only way I see this can be done is by closing/flushing all `Scribes` and recreating the same new ones with juts the `Severity` different.

An easier way could be to implement `permitItem` as a class method
```permitItem :: ScribeName -> Severity -> Item a -> IO Bool``` (maybe in `Katip m`)
which would allow dynamically changing `Severity` as read from `IORef`.

What do you think? Or am I missing something obvious?

**Comment by MichaelXavier** [comment]
I don't think a class method on `Katip` is the right way to go. The `Katip` monad just says how to get at your log env in order to do logging. Having a logical class method on it to do log filtering doesn't make a lot of sense to me because it is sort of an application level concern that can even change dynamically as you're trying to do, and typeclasses are meant to be universal things and are notoriously hard to customize.

A `Scribe` looks like this:
```haskell
data Scribe = Scribe {
     liPush          :: forall a. LogItem a => Item a -> IO ()
   , scribeFinalizer :: IO ()
   }
```

A `LogEnv` has a `M.Map Text ScribeHandle` and a `ScribeHandle` look like this:
```haskell
data ScribeHandle = ScribeHandle {
      shScribe :: Scribe
    , shChan :: BQ.TBQueue WorkerMessage
    }
```

We publish a standard `permitItem` combinator that scribes may use to determine if they should log an item or not. It is pretty ad hoc and I'm open to changing it. If you're using a standard scribe and want to dynamically change the permissiveness of it, you are right in that the filtering aspect is pushed deep into the `Scribe` initialization and is thus not easy to change, or rather impossible to make *more* permissive.

So we can think of how we can push this feature higher up the stack to open it to customization. We must have per-scribe level granularity because that's what we have to day and it makes sense to let different scribes log at different severities. Also, we could take this opportunity to not shackle ourselves to using only severity to filter messages. Perhaps there's an application-specific scribe that should not log things from a certain namespace. 

My initial thought is to add a field to `Scribe` that contains the logic for permitting items. That function would be consulted on push. A side benefit here: unlike today, if an item would not be permitted for a scribe, it would not even be buffered into the queue and held in memory but would be discarded immediately.  Thus, if your program has an extremely high volume `DebugS` log message but is configured for `InfoS` level logging, right now those `DebugS` messages go into the buffer and take up space, rather than being discarded immediately.

So I'd propose something like:
```haskell
data Scribe = Scribe {
     liPush          :: forall a. LogItem a => Item a -> IO ()
   , scribeFinalizer :: IO ()
   , scribePermitItem :: Item a => IO Bool
   }
```
Existing scribes would default to `pure . permitItem` for backwards compatibility. A dynamic `LogEnv` could either modify the `Scribe` or do a reader-style `local` with the existing machinery to temporarily modify the filtering criteria for a span of code. Thoughts? Tagging in @ozataman

**Comment by vlatkoB** [comment]
On first read this sounds good, especially because we get easy `local` with this approach.
Have I understood right that `scribePermitItem` will be called by `Katip`, not by `Scribe`, so scribes can't ignore it like today?
(small correction: the "fat" arrow in `scribePermitItem` should be "thin".)

**Comment by MichaelXavier** [comment]
Good catch on the arrow. Edited that. `katip` the library would internally call `scribePermitItem` to decide whether or not to send the item. This has another benefit in that right now scribe authors must remember to do item filtering internally and they would no longer have to. It would probably be best to make this a breaking change and actually rename `permitItem` to something like `defaultPermitItem` perhaps to clarify that it is going to go in a field of scribe rather than within the "guts" of the scribe itself. There should probably be documentation to the effect that scribes should always be constructed with `defaultPermitItem` by libraries and that it is meant for user-land extension.

**Comment by vlatkoB** [comment]
Maybe just hide `Scribe` constructor and expose only `mkScribe` and `Scribe`'s lenses?

**Comment by vlatkoB** [comment]
If `scribePermitItem` would be on the higher level, then the old `permitItem` could emit deprecation warning and be changed internally to no-op. So even with filtering in custom scribes, nothing will be changed and no performance hit.

**Comment by vlatkoB** [comment]
Can you take a look at [this first try](https://bitbucket.org/vlatkoB/katip/commits/d0290c819c8079eaf071ff5df42e6726bfd0e5e0) to see if I found the best place to integrate `scribePermitItem` filtering, and the TODO @574?

**Comment by vlatkoB** [comment]
I found some time to mostly finish this. You can see it at https://bitbucket.org/vlatkoB/katip/src/master/.
I also removed Severity from mk*Scribe functions as they are not needed anymore.
Tested with 
```onlyDebug :: Item a -> IO Bool
onlyDebug Item{..} = return $ _itemSeverity == DebugS
                           || "ALL" `elem` unNamespace _itemNamespace
```
and these scribes
```
mkHandleScribe ColorIfTerminal stdout V3 onlyDebug
mkFileScribe fNm V3 permitAll
```
and output is what is expected to be. Works both for scribe list and `mappend`ed scribes.
Comments?

**Comment by MichaelXavier** [comment]
Sorry for the delay, its been hard to set aside time for this. It is a bit hard to get a full view of the commit, so a PR would be great. It does seem like there's a lot of disruption going on in the APIs. I think at the very least we should be exporting a `PermitFunc` that looks at severity so that it is easy for users to replicate the existing, common behavior of filtering at a severity level. I also noticed that `ioLogEnv` has completely shed the severity filter and will now log everything, which doesn't seem quite right. I think we need to establish a pretty consistent convention in the core library of how scribes should look, so maybe they always take a `PermitFunc`.

So maybe if you think this is ready and you fix that one minor issue with `ioLogEnv`, you could submit this as a PR and I could see what some of my colleagues think. Thanks again for getting all this together!

**Comment by vlatkoB** [comment]
I [squashed](https://bitbucket.org/vlatkoB/katip/commits/dd4d56e9c308b6956b2c0f58a69634243add03dd) all commits and now all the changes are easier to track. Not so many API changes.
I reused `permitItem` and `ioLogEnv` is as described above.

Please take a look and let me know if there is anything else you'd like to discuss/change before I make the PR, so others can do the final review and test.

BTW, should the version of `katip-elasticsearch` be unified with version of `katip`?

**Comment by MichaelXavier** [comment]
The PR looks much clearer, thank you! I think everything looks good. I am in agreement that we can break the API of `ioLogEnv` for the sake of having a uniform convention that a scribe constructor takes a permit function. `katip-elasticsearch` is sort of like the canonical second scribe after the handle scribe, so it is good to keep that in sync.

I think this PR is ready for submission :smiley:

**Comment by vlatkoB** [comment]
PR is done. Sorry for some noise on Travis. :-(
I moved the fork to [GitHub](https://github.com/vlatkoB/katip)
